lib/interceptor/decompress.js

```diff
@@ -0,0 +1,131 @@
+
+const nullBodyStatus = [101, 204, 205, 304]
+
+class DecompressHandler extends DecoratorHandler {
+  #handler
+  #opts
+
+  #inputStream = null
+  #trailers = null
+
+  constructor (opts, handler) {
+    super(handler)
+    this.#handler = handler
+    this.#opts = opts
+  }
+
+  onConnect (abort) {
+    this.#inputStream = null
+    this.#trailers = null
+    return this.#handler.onConnect(abort)
+  }
+
+  #onDecompressStreamFinished (err) {
+    if (err) {
+      this.#handler.onError(err)
+    } else {
+      this.#handler.onComplete(this.#trailers)
+    }
+  }
+
+  onHeaders (statusCode, rawHeaders, resume, statusMessage) {
+    const parsedHeaders = util.parseHeaders(rawHeaders)
+    const contentEncoding = parsedHeaders['content-encoding']
+    const requestEncodings = contentEncoding ? contentEncoding.split(',').map(e => e.trim().toLowerCase()) : []
+
+    const { method } = this.#opts
+
+    // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding
+    if (requestEncodings.length !== 0 && method !== 'HEAD' && method !== 'CONNECT' && !nullBodyStatus.includes(statusCode)) {
+      const decoders = []
+      for (let i = 0; i < requestEncodings.length; ++i) {
+        const requestEncoding = requestEncodings[i]
+        // https://www.rfc-editor.org/rfc/rfc9112.html#section-7.2
+        if ((requestEncoding === 'x-gzip' || requestEncoding === 'gzip')) {
+          decoders.push(zlib.createGunzip({
+            // Be less strict when decoding compressed responses, since sometimes
+            // servers send slightly invalid responses that are still accepted
+            // by common browsers.
+            // Always using Z_SYNC_FLUSH is what cURL does.
+            flush: zlib.constants.Z_SYNC_FLUSH,
+            finishFlush: zlib.constants.Z_SYNC_FLUSH
+          }))
+        } else if (requestEncoding === 'deflate') {
+          decoders.push(createInflate())
+        } else if (requestEncoding === 'br') {
+          decoders.push(zlib.createBrotliDecompress())
+        } else {
+          decoders.length = 0
+          break
+        }
+      }
+
+      if (decoders.length !== 0) {
+        const [firstDecoder, ...restDecoders] = decoders
+        this.#inputStream = firstDecoder
+        let outputStream = firstDecoder
+
+        if (restDecoders.length !== 0) {
+          outputStream = pipeline(
+            this.#inputStream,
+            ...restDecoders,
+            this.#onDecompressStreamFinished.bind(this)
+          )
+        } else {
+          finished(
+            this.#inputStream,
+            this.#onDecompressStreamFinished.bind(this)
+          )
+        }
+
+        outputStream.on('data', (chunk) => {
+          if (!this.#handler.onData(chunk)) {
+            this.#inputStream.pause()
+            this.#inputStream.once('drain', () => this.#inputStream.resume())
+          }
+        })
+      }
+    }
+
+    return this.#handler.onHeaders(
```

Comment:
Thanks for putting up this pr! I added this interceptor into my project. 

This interceptor could also be used with fetch if you remove the `content-encoding`. My use case is I have a LoggingInterceptor that sends the response body to GCP logging. This worked great except one API sends a gzipped response. Now I have the following flow:

Decompress -> Logging -> Fetch

```javascript
    delete parsedHeaders['content-encoding'];
    const newRawHeaders = Object.entries(parsedHeaders)
      .flat()
      .map((e) => Buffer.from(e));

    return this.#handler.onHeaders(statusCode, newRawHeaders, resume, statusMessage);
```

Reply:
Welp - I ended up having issues with the interceptor. Haven't been able to dive in yet, but I feel like my flow should work. Will try again once this is reviewed/merged.